README.md
# yyl-ssr-fetcher

用于 服务端 请求

## 安装

```bash
# yarn
yarn add yyl-ssr-fetcher

# npm
npm i yyl-ssr-fetcher --save
```

## usage

```typescript
import { CoreOptions } from 'request'
/** 日志类型 */
export declare enum LogType {
  Info = 'info',
  Success = 'success',
  Warn = 'warn',
  Error = 'error'
}
/** 日志参数 */
export interface LogProps {
  /** 日志类型 */
  type: LogType
  /** 日志来源 */
  path: string
  /** 日志内容 */
  args: any[]
}
/** request 配置 */
export interface RequestOption extends CoreOptions {
  /** 请求来源-用于logger */
  ref?: string
}
/** fetcher 配置 */
export interface FetcherOption {
  /** 超时设置 */
  timeout?: number
  /** 超时设置 */
  userAgent?: string
  /** request 配置 */
  requestOption?: RequestOption
  /** 默认请求协议 */
  defaultProtocol?: string
  /** 日志 */
  logger?: (props: LogProps) => void
}
/** fetcher 属性 */
export declare type FetcherProperty = Required<FetcherOption>
/** fetcher 主函数 */
export declare class Fetcher {
  /** 超时时间 */
  private timeout
  /** 请求ua */
  private userAgent
  /** request 配置 */
  private requestOption
  /** 默认 protocol */
  private defaultProtocol
  /** 默认logger */
  private logger
  constructor(option?: FetcherOption)
  /** url 格式化 */
  private formatUrl
  /** get 请求 */
  get<I = {}, O = {}>(url: string, req?: I, option?: FetcherOption['requestOption']): Promise<O>
  /** post 请求 */
  post<I = {}, O = {}>(url: string, req?: I, option?: FetcherOption['requestOption']): Promise<O>
}
```


history.md
# 版本历史

## 0.2.3 (2020-12-29)

- feat: 补充 请求耗时 log

## 0.2.1 (2020-11-24)

- feat: bugfix

## 0.2.0 (2020-11-24)

- feat: 新增 logger 参数

## 0.1.2 (2020-11-17)

- feat: output 文件不再压缩

## 0.1.1 (2020-11-17)

- fix: 修复 url 没有 处理 protocol 问题

## 0.1.0 (2020-11-16)

- feat: 诞生


test/jest/index.test.js
const { Fetcher } = require('../../')
test('get usage test', async () => {
  const fetcher = new Fetcher({ timeout: 2000 })

  const rs = await fetcher.get('https://www.yy.com/yyweb/module/data/header')
  expect(rs).not.toEqual({})
})

test('post usage test', async () => {
  const fetcher = new Fetcher({ timeout: 2000 })

  const rs = await fetcher.post('https://www.yy.com/yyweb/module/data/header')
  expect(rs).not.toEqual({})
})


output/index.js
/*!
 * yyl-ssr-fetcher cjs 0.2.3
 * (c) 2020 - 2020 jackness
 * Released under the MIT License.
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var request = _interopDefault(require('request'));
var extend = _interopDefault(require('extend'));

(function (LogType) {
    LogType["Info"] = "info";
    LogType["Success"] = "success";
    LogType["Warn"] = "warn";
    LogType["Error"] = "error";
})(exports.LogType || (exports.LogType = {}));
/** 无协议匹配 */
const NO_PROTOCOL_REG = /^\/\//;
/** 默认ua */
const DEFAULT_UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.122 Safari/537.36 ServerSide';
/** fetcher 主函数 */
class Fetcher {
    constructor(option) {
        /** 超时时间 */
        this.timeout = 5000;
        /** 请求ua */
        this.userAgent = DEFAULT_UA;
        /** request 配置 */
        this.requestOption = {};
        /** 默认 protocol */
        this.defaultProtocol = 'http:';
        /** 默认logger */
        this.logger = () => { };
        if (option === null || option === void 0 ? void 0 : option.timeout) {
            this.timeout = option.timeout;
        }
        if (option === null || option === void 0 ? void 0 : option.userAgent) {
            this.userAgent = option.userAgent;
        }
        if (option === null || option === void 0 ? void 0 : option.requestOption) {
            this.requestOption = option.requestOption;
        }
        // request option 配置
        this.requestOption = extend({
            headers: {
                'User-Agent': this.userAgent
            },
            timeout: this.timeout
        }, this.requestOption);
        // logger 初始化
        if (option === null || option === void 0 ? void 0 : option.logger) {
            this.logger = option.logger;
        }
    }
    /** url 格式化 */
    formatUrl(url) {
        const { defaultProtocol } = this;
        let rUrl = url;
        if (rUrl.match(NO_PROTOCOL_REG)) {
            rUrl = `${defaultProtocol}${rUrl}`;
        }
        return rUrl;
    }
    /** get 请求 */
    get(url, req, option) {
        const start = +new Date();
        const { requestOption } = this;
        const rUrl = this.formatUrl(url);
        const logPrefix = `请求${rUrl}`;
        const ref = (option === null || option === void 0 ? void 0 : option.ref) || rUrl;
        let param = Object.assign({ qs: req || {} }, requestOption);
        if (requestOption) {
            param = extend(param, option);
        }
        return new Promise((resolve, reject) => {
            this.logger({
                path: ref,
                type: exports.LogType.Info,
                args: [logPrefix, '类型 GET', '参数', param]
            });
            request.get(rUrl, param, (err, res, body) => {
                const cost = +new Date() - start;
                if (!err) {
                    if (res.statusCode === 200) {
                        try {
                            const r = JSON.parse(body);
                            this.logger({
                                path: ref,
                                type: exports.LogType.Success,
                                args: [`${logPrefix} 成功(${cost}ms)`, '返回值', r]
                            });
                            resolve(r);
                        }
                        catch (er) {
                            const errMsg = `parse error: ${body}`;
                            this.logger({
                                path: ref,
                                type: exports.LogType.Error,
                                args: [`${logPrefix} 失败`, errMsg]
                            });
                            reject(new Error(`${logPrefix}失败: ${errMsg}`));
                        }
                    }
                    else {
                        const errMsg = `状态非 200: ${res.statusCode}`;
                        this.logger({
                            path: ref,
                            type: exports.LogType.Error,
                            args: [`${logPrefix} 失败`, errMsg]
                        });
                        reject(new Error(`${logPrefix}失败: ${errMsg}`));
                    }
                }
                else {
                    this.logger({
                        path: ref,
                        type: exports.LogType.Error,
                        args: [`${logPrefix} 失败`, err]
                    });
                    reject(err);
                }
            });
        });
    }
    /** post 请求 */
    post(url, req, option) {
        const start = +new Date();
        const { requestOption } = this;
        const rUrl = this.formatUrl(url);
        const logPrefix = `请求${rUrl}`;
        const ref = (option === null || option === void 0 ? void 0 : option.ref) || rUrl;
        let param = Object.assign({ formData: req || {} }, requestOption);
        if (requestOption) {
            param = extend(param, option);
        }
        return new Promise((resolve, reject) => {
            this.logger({
                path: ref,
                type: exports.LogType.Info,
                args: [logPrefix, '类型 Post', '参数', param]
            });
            request.post(rUrl, param, (err, res, body) => {
                const cost = +new Date() - start;
                if (!err) {
                    if (res.statusCode === 200) {
                        try {
                            const r = JSON.parse(body);
                            this.logger({
                                path: ref,
                                type: exports.LogType.Success,
                                args: [`${logPrefix} 成功(${cost})ms`, '返回值', r]
                            });
                            resolve(r);
                        }
                        catch (er) {
                            const errMsg = `parse error: ${body}`;
                            this.logger({
                                path: ref,
                                type: exports.LogType.Error,
                                args: [`${logPrefix} 失败`, errMsg]
                            });
                            reject(new Error(`${logPrefix}失败: ${errMsg}`));
                        }
                    }
                    else {
                        const errMsg = `状态非 200: ${res.statusCode}`;
                        this.logger({
                            path: ref,
                            type: exports.LogType.Error,
                            args: [`${logPrefix} 失败`, errMsg]
                        });
                        reject(new Error(`${logPrefix}失败: ${errMsg}`));
                    }
                }
                else {
                    this.logger({
                        path: ref,
                        type: exports.LogType.Error,
                        args: [`${logPrefix} 失败`, err]
                    });
                    reject(err);
                }
            });
        });
    }
}

exports.Fetcher = Fetcher;


rollup.config.js
import pkg from './package.json'
import typescript from 'rollup-plugin-typescript2'
import commonjs from 'rollup-plugin-commonjs'
import nodeResolve from 'rollup-plugin-node-resolve'
import json from 'rollup-plugin-json'

function buildBanner(type) {
  return [
    '/*!',
    ` * ${pkg.name} ${type} ${pkg.version}`,
    ` * (c) 2020 - ${new Date().getFullYear()} jackness`,
    ' * Released under the MIT License.',
    ' */'
  ].join('\n')
}

const IS_PUBLISH = process.env.NODE_ENV === 'production'

const config = {
  input: './src/index.ts',
  output: [],
  plugins: [
    nodeResolve({ jsnext: true }),
    commonjs(),
    json(),
    typescript({
      typescript: require('typescript')
    })
  ],
  external: ['request', 'extend']
}

export default [
  {
    input: config.input,
    output: [
      {
        file: './output/index.js',
        format: 'cjs',
        banner: buildBanner('cjs'),
        exports: 'named',
        sourcemap: false
      }
    ],
    plugins: config.plugins,
    external: config.external
  },
  {
    input: config.input,
    output: [
      {
        file: './output/index.esm.js',
        format: 'esm',
        banner: buildBanner('esm'),
        sourcemap: false
      }
    ],
    plugins: config.plugins,
    external: config.external
  }
]


output/index.d.ts
import { CoreOptions } from 'request';
/** 日志类型 */
export declare enum LogType {
    Info = "info",
    Success = "success",
    Warn = "warn",
    Error = "error"
}
/** 日志参数 */
export interface LogProps {
    /** 日志类型 */
    type: LogType;
    /** 日志来源 */
    path: string;
    /** 日志内容 */
    args: any[];
}
/** request 配置 */
export interface RequestOption extends CoreOptions {
    /** 请求来源-用于logger */
    ref?: string;
}
/** fetcher 配置 */
export interface FetcherOption {
    /** 超时设置 */
    timeout?: number;
    /** 超时设置 */
    userAgent?: string;
    /** request 配置 */
    requestOption?: RequestOption;
    /** 默认请求协议 */
    defaultProtocol?: string;
    /** 日志 */
    logger?: (props: LogProps) => void;
}
/** fetcher 属性 */
export declare type FetcherProperty = Required<FetcherOption>;
/** fetcher 主函数 */
export declare class Fetcher {
    /** 超时时间 */
    private timeout;
    /** 请求ua */
    private userAgent;
    /** request 配置 */
    private requestOption;
    /** 默认 protocol */
    private defaultProtocol;
    /** 默认logger */
    private logger;
    constructor(option?: FetcherOption);
    /** url 格式化 */
    private formatUrl;
    /** get 请求 */
    get<I = {}, O = {}>(url: string, req?: I, option?: FetcherOption['requestOption']): Promise<O>;
    /** post 请求 */
    post<I = {}, O = {}>(url: string, req?: I, option?: FetcherOption['requestOption']): Promise<O>;
}


output/index.esm.js
/*!
 * yyl-ssr-fetcher esm 0.2.3
 * (c) 2020 - 2020 jackness
 * Released under the MIT License.
 */
import request from 'request';
import extend from 'extend';

/** 日志类型 */
var LogType;
(function (LogType) {
    LogType["Info"] = "info";
    LogType["Success"] = "success";
    LogType["Warn"] = "warn";
    LogType["Error"] = "error";
})(LogType || (LogType = {}));
/** 无协议匹配 */
const NO_PROTOCOL_REG = /^\/\//;
/** 默认ua */
const DEFAULT_UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.122 Safari/537.36 ServerSide';
/** fetcher 主函数 */
class Fetcher {
    constructor(option) {
        /** 超时时间 */
        this.timeout = 5000;
        /** 请求ua */
        this.userAgent = DEFAULT_UA;
        /** request 配置 */
        this.requestOption = {};
        /** 默认 protocol */
        this.defaultProtocol = 'http:';
        /** 默认logger */
        this.logger = () => { };
        if (option === null || option === void 0 ? void 0 : option.timeout) {
            this.timeout = option.timeout;
        }
        if (option === null || option === void 0 ? void 0 : option.userAgent) {
            this.userAgent = option.userAgent;
        }
        if (option === null || option === void 0 ? void 0 : option.requestOption) {
            this.requestOption = option.requestOption;
        }
        // request option 配置
        this.requestOption = extend({
            headers: {
                'User-Agent': this.userAgent
            },
            timeout: this.timeout
        }, this.requestOption);
        // logger 初始化
        if (option === null || option === void 0 ? void 0 : option.logger) {
            this.logger = option.logger;
        }
    }
    /** url 格式化 */
    formatUrl(url) {
        const { defaultProtocol } = this;
        let rUrl = url;
        if (rUrl.match(NO_PROTOCOL_REG)) {
            rUrl = `${defaultProtocol}${rUrl}`;
        }
        return rUrl;
    }
    /** get 请求 */
    get(url, req, option) {
        const start = +new Date();
        const { requestOption } = this;
        const rUrl = this.formatUrl(url);
        const logPrefix = `请求${rUrl}`;
        const ref = (option === null || option === void 0 ? void 0 : option.ref) || rUrl;
        let param = Object.assign({ qs: req || {} }, requestOption);
        if (requestOption) {
            param = extend(param, option);
        }
        return new Promise((resolve, reject) => {
            this.logger({
                path: ref,
                type: LogType.Info,
                args: [logPrefix, '类型 GET', '参数', param]
            });
            request.get(rUrl, param, (err, res, body) => {
                const cost = +new Date() - start;
                if (!err) {
                    if (res.statusCode === 200) {
                        try {
                            const r = JSON.parse(body);
                            this.logger({
                                path: ref,
                                type: LogType.Success,
                                args: [`${logPrefix} 成功(${cost}ms)`, '返回值', r]
                            });
                            resolve(r);
                        }
                        catch (er) {
                            const errMsg = `parse error: ${body}`;
                            this.logger({
                                path: ref,
                                type: LogType.Error,
                                args: [`${logPrefix} 失败`, errMsg]
                            });
                            reject(new Error(`${logPrefix}失败: ${errMsg}`));
                        }
                    }
                    else {
                        const errMsg = `状态非 200: ${res.statusCode}`;
                        this.logger({
                            path: ref,
                            type: LogType.Error,
                            args: [`${logPrefix} 失败`, errMsg]
                        });
                        reject(new Error(`${logPrefix}失败: ${errMsg}`));
                    }
                }
                else {
                    this.logger({
                        path: ref,
                        type: LogType.Error,
                        args: [`${logPrefix} 失败`, err]
                    });
                    reject(err);
                }
            });
        });
    }
    /** post 请求 */
    post(url, req, option) {
        const start = +new Date();
        const { requestOption } = this;
        const rUrl = this.formatUrl(url);
        const logPrefix = `请求${rUrl}`;
        const ref = (option === null || option === void 0 ? void 0 : option.ref) || rUrl;
        let param = Object.assign({ formData: req || {} }, requestOption);
        if (requestOption) {
            param = extend(param, option);
        }
        return new Promise((resolve, reject) => {
            this.logger({
                path: ref,
                type: LogType.Info,
                args: [logPrefix, '类型 Post', '参数', param]
            });
            request.post(rUrl, param, (err, res, body) => {
                const cost = +new Date() - start;
                if (!err) {
                    if (res.statusCode === 200) {
                        try {
                            const r = JSON.parse(body);
                            this.logger({
                                path: ref,
                                type: LogType.Success,
                                args: [`${logPrefix} 成功(${cost})ms`, '返回值', r]
                            });
                            resolve(r);
                        }
                        catch (er) {
                            const errMsg = `parse error: ${body}`;
                            this.logger({
                                path: ref,
                                type: LogType.Error,
                                args: [`${logPrefix} 失败`, errMsg]
                            });
                            reject(new Error(`${logPrefix}失败: ${errMsg}`));
                        }
                    }
                    else {
                        const errMsg = `状态非 200: ${res.statusCode}`;
                        this.logger({
                            path: ref,
                            type: LogType.Error,
                            args: [`${logPrefix} 失败`, errMsg]
                        });
                        reject(new Error(`${logPrefix}失败: ${errMsg}`));
                    }
                }
                else {
                    this.logger({
                        path: ref,
                        type: LogType.Error,
                        args: [`${logPrefix} 失败`, err]
                    });
                    reject(err);
                }
            });
        });
    }
}

export { Fetcher, LogType };


src/index.ts
import request, { CoreOptions } from 'request'
import extend from 'extend'

/** 日志类型 */
export enum LogType {
  Info = 'info',
  Success = 'success',
  Warn = 'warn',
  Error = 'error'
}

/** 日志参数 */
export interface LogProps {
  /** 日志类型 */
  type: LogType
  /** 日志来源 */
  path: string
  /** 日志内容 */
  args: any[]
}

/** request 配置 */
export interface RequestOption extends CoreOptions {
  /** 请求来源-用于logger */
  ref?: string
}
/** fetcher 配置 */
export interface FetcherOption {
  /** 超时设置 */
  timeout?: number
  /** 超时设置 */
  userAgent?: string
  /** request 配置 */
  requestOption?: RequestOption
  /** 默认请求协议 */
  defaultProtocol?: string
  /** 日志 */
  logger?: (props: LogProps) => void
}

/** 无协议匹配 */
const NO_PROTOCOL_REG = /^\/\//

/** fetcher 属性 */
export type FetcherProperty = Required<FetcherOption>

/** 默认ua */
const DEFAULT_UA =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.122 Safari/537.36 ServerSide'

/** fetcher 主函数 */
export class Fetcher {
  /** 超时时间 */
  private timeout: FetcherProperty['timeout'] = 5000
  /** 请求ua */
  private userAgent: FetcherProperty['userAgent'] = DEFAULT_UA
  /** request 配置 */
  private requestOption: FetcherProperty['requestOption'] = {}
  /** 默认 protocol */
  private defaultProtocol: FetcherProperty['defaultProtocol'] = 'http:'
  /** 默认logger */
  private logger: FetcherProperty['logger'] = () => {}
  constructor(option?: FetcherOption) {
    if (option?.timeout) {
      this.timeout = option.timeout
    }
    if (option?.userAgent) {
      this.userAgent = option.userAgent
    }

    if (option?.requestOption) {
      this.requestOption = option.requestOption
    }

    // request option 配置
    this.requestOption = extend(
      {
        headers: {
          'User-Agent': this.userAgent
        },
        timeout: this.timeout
      },
      this.requestOption
    )

    // logger 初始化
    if (option?.logger) {
      this.logger = option.logger
    }
  }

  /** url 格式化 */
  private formatUrl(url: string) {
    const { defaultProtocol } = this
    let rUrl = url
    if (rUrl.match(NO_PROTOCOL_REG)) {
      rUrl = `${defaultProtocol}${rUrl}`
    }
    return rUrl
  }

  /** get 请求 */
  get<I = {}, O = {}>(url: string, req?: I, option?: FetcherOption['requestOption']) {
    const start = +new Date()
    const { requestOption } = this
    const rUrl = this.formatUrl(url)
    const logPrefix = `请求${rUrl}`
    const ref = option?.ref || rUrl

    let param: FetcherProperty['requestOption'] = {
      qs: req || {},
      ...requestOption
    }
    if (requestOption) {
      param = extend<FetcherProperty['requestOption'], FetcherOption['requestOption']>(
        param,
        option
      )
    }

    return new Promise<O>((resolve, reject) => {
      this.logger({
        path: ref,
        type: LogType.Info,
        args: [logPrefix, '类型 GET', '参数', param]
      })
      request.get(rUrl, param, (err, res, body) => {
        const cost = +new Date() - start
        if (!err) {
          if (res.statusCode === 200) {
            try {
              const r = JSON.parse(body)
              this.logger({
                path: ref,
                type: LogType.Success,
                args: [`${logPrefix} 成功(${cost}ms)`, '返回值', r]
              })
              resolve(r)
            } catch (er) {
              const errMsg = `parse error: ${body}`
              this.logger({
                path: ref,
                type: LogType.Error,
                args: [`${logPrefix} 失败`, errMsg]
              })
              reject(new Error(`${logPrefix}失败: ${errMsg}`))
            }
          } else {
            const errMsg = `状态非 200: ${res.statusCode}`
            this.logger({
              path: ref,
              type: LogType.Error,
              args: [`${logPrefix} 失败`, errMsg]
            })
            reject(new Error(`${logPrefix}失败: ${errMsg}`))
          }
        } else {
          this.logger({
            path: ref,
            type: LogType.Error,
            args: [`${logPrefix} 失败`, err]
          })
          reject(err)
        }
      })
    })
  }

  /** post 请求 */
  post<I = {}, O = {}>(url: string, req?: I, option?: FetcherOption['requestOption']) {
    const start = +new Date()
    const { requestOption } = this
    const rUrl = this.formatUrl(url)
    const logPrefix = `请求${rUrl}`
    const ref = option?.ref || rUrl

    let param: FetcherProperty['requestOption'] = {
      formData: req || {},
      ...requestOption
    }
    if (requestOption) {
      param = extend<FetcherProperty['requestOption'], FetcherOption['requestOption']>(
        param,
        option
      )
    }

    return new Promise<O>((resolve, reject) => {
      this.logger({
        path: ref,
        type: LogType.Info,
        args: [logPrefix, '类型 Post', '参数', param]
      })

      request.post(rUrl, param, (err, res, body) => {
        const cost = +new Date() - start
        if (!err) {
          if (res.statusCode === 200) {
            try {
              const r = JSON.parse(body)
              this.logger({
                path: ref,
                type: LogType.Success,
                args: [`${logPrefix} 成功(${cost})ms`, '返回值', r]
              })
              resolve(r)
            } catch (er) {
              const errMsg = `parse error: ${body}`
              this.logger({
                path: ref,
                type: LogType.Error,
                args: [`${logPrefix} 失败`, errMsg]
              })
              reject(new Error(`${logPrefix}失败: ${errMsg}`))
            }
          } else {
            const errMsg = `状态非 200: ${res.statusCode}`
            this.logger({
              path: ref,
              type: LogType.Error,
              args: [`${logPrefix} 失败`, errMsg]
            })
            reject(new Error(`${logPrefix}失败: ${errMsg}`))
          }
        } else {
          this.logger({
            path: ref,
            type: LogType.Error,
            args: [`${logPrefix} 失败`, err]
          })
          reject(err)
        }
      })
    })
  }
}
